Android support for non-rooted phones

TroyNeubauer opened the pull request:
Adds a `Device::from_fd` function for wrapping devices obtained from android's [UsbManager](https://developer.android.com/reference/android/hardware/usb/UsbManager)

### Testing

Tested with HackrfOne (firmware version `2024.02.1`) on an A7 lite running Android 14 (kernel version `4.19.191`).

The basic configuration methods work, as well as receiving from bulk endpoints via the async Queue interface: 
See: https://github.com/MerchGuardian/seify/blob/hackrf/crates/hackrf-one-rs/src/lib.rs#L535-L566

```
let dev = nusb::Device::from_fd(fd).unwrap();

for config in dev.configurations() {
    log::info!("Got config: {:?}", config);

    log::info!("num_interfaces: {:?}", config.num_interfaces());
    log::info!("configuration_value: {:?}", config.configuration_value());

    log::info!("attributes: {:?}", config.attributes());
    log::info!("max_power: {:?}", config.max_power());
    let string_index = config.string_index().unwrap();
    log::info!("string_index: {:?}", string_index);

    for lang in dev.get_string_descriptor_supported_languages(Duration::from_millis(500)).unwrap() {
        log::info!("Got lang: {lang:?}");
        let s = dev.get_string_descriptor(string_index, lang, Duration::from_millis(500)).unwrap();
        log::info!("str descriptor: {s:?}");
    }
}
```

Prints:

```
INFO:android_test - Got config: Configuration {
    configuration_value: 1,
    num_interfaces: 1,
    attributes: 128,
    max_power: 250,
    string_index: Some(
        3,
    ),
    interface_alt_settings: [
        InterfaceAltSetting {
            interface_number: 0,
            alternate_setting: 0,
            num_endpoints: 2,
            class: 255,
            subclass: 255,
            protocol: 255,
            string_index: None,
            endpoints: [
                Endpoint {
                    address: 0x81,
                    direction: In,
                    transfer_type: Bulk,
                    max_packet_size: 512,
                    packets_per_microframe: 1,
                    interval: 0,
                },
                Endpoint {
                    address: 0x02,
                    direction: Out,
                    transfer_type: Bulk,
                    max_packet_size: 512,
                    packets_per_microframe: 1,
                    interval: 0,
                },
            ],
        },
    ],
}
INFO:android_test - num_interfaces: 1
INFO:android_test - configuration_value: 1
INFO:android_test - attributes: 128
INFO:android_test - max_power: 250
INFO:android_test - string_index: 3
INFO:android_test - Got lang: 1033
INFO:android_test - str descriptor: "Transceiver"
```

TODO:
* Add CI job that builds with `cargo ndk` to ensure we hit the new `#[cfg(target_os = "android")] blocks in CI

kevinmehall commented on src/platform/linux_usbfs/device.rs:
```diff
@@ -54,8 +56,105 @@ impl LinuxDevice {
+        debug!("Wrapping fd {} as usbfs device", fd.as_raw_fd());
+
+        // NOTE: must be called from the same thread that gave us the FD on android, otherwise this fails
+        let proc_path = PathBuf::from(format!("/proc/self/fd/{}", fd.as_raw_fd()));
+
+        debug!("Reading link: {proc_path:?}");
+        let fd_path = std::fs::read_link(&proc_path)?;
+
+        // TODO: should we switch to regex or regex-lite for this?
+        let prefix = "/dev/bus/usb/";
+        let Ok(dev_num_bus_num) = fd_path.strip_prefix(prefix) else {
+            error!(
+                "Usb device path does not start with required prefix `{prefix}`: {}",
+                fd_path.to_string_lossy()
+            );
+            return Err(ErrorKind::Other.into());
+        };
+        let Some(dev_num_bus_num) = dev_num_bus_num.to_str() else {
+            error!("End of usb device path is not UTF-8!: {dev_num_bus_num:?}",);
+            return Err(ErrorKind::Other.into());
+        };
+
+        let mut split = dev_num_bus_num.split("/");
+        let Some(bus_str) = split.next() else {
+            error!("Failed to split usbfs device path: {}", dev_num_bus_num);
+            return Err(ErrorKind::Other.into());
+        };
+        let Some(dev_str) = split.next() else {
+            error!("Failed to split usbfs device path: {}", dev_num_bus_num);
+            return Err(ErrorKind::Other.into());
+        };
+
+        let Ok(busnum) = bus_str.parse::<u8>() else {
+            error!("Usb bus number failed to parse: {bus_str}",);
+            return Err(ErrorKind::Other.into());
+        };
+        let Ok(devnum) = dev_str.parse::<u8>() else {
+            error!("Usb device number failed to parse: {dev_str}",);
+            return Err(ErrorKind::Other.into());
+        };
```
`create_inner` doesn't do anything with `busnum` and `devnum` beyond logging it. I wouldn't mind changing that log message to avoid going so far out of your way to get these.

TroyNeubauer replied:
Ah, good catch. Got caught up trying to replicate the libusb code

kevinmehall commented on src/platform/linux_usbfs/device.rs:
```diff
@@ -54,8 +56,105 @@ impl LinuxDevice {
+            return Err(ErrorKind::Other.into());
+        };
+
+        const LIBUSB_REQUEST_GET_CONFIGURATION: u8 = 0x08;
```
```suggestion
        const REQUEST_GET_CONFIGURATION: u8 = 0x08;
```

(it's a request defined by the USB standard, not something from libusb)

kevinmehall commented on src/platform/linux_usbfs/device.rs:
```diff
@@ -54,8 +56,105 @@ impl LinuxDevice {
+            index: 0,
+        };
+
+        // TODO: what to do about blocking here?
```
Opening the device file also blocks for USB resume and a `GET_STATUS` request, so I guess it's no worse than the other way to open a device?

TroyNeubauer replied:
Makes sense. I suppose this asks a larger question about blocking since its likely that the user will call one of these methods on an async runtime if they later use the async transfer methods

kevinmehall commented on src/platform/linux_usbfs/device.rs:
```diff
@@ -54,8 +56,105 @@ impl LinuxDevice {
+        // See: https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/os/linux_usbfs.c#L865
+        let n = r.map_err(errno_to_transfer_error)?;
+        if n != dst.len() {
+            error!("Failed to read descriptor");
```
```suggestion
            error!("GET_CONFIGURATION request failed");
```
Since it's not a descriptor

kevinmehall commented on src/platform/linux_usbfs/device.rs:
```diff
@@ -54,8 +56,105 @@ impl LinuxDevice {
+            },
+        );
+
+        // TODO: Could add some more logic here to support buggy devices, similar to what libusb does
+        // See: https://github.com/libusb/libusb/blob/467b6a8896daea3d104958bf0887312c5d14d150/libusb/os/linux_usbfs.c#L865
```
I wouldn't be surprised if it's common to not implement `GET_CONFIGURATION` because it doesn't seem like it would be commonly used.

If there's only one configuration, which will be the case for the vast majority of devices, it seems pretty safe to assume that's the active configuration.

kevinmehall reviewed (commented):
Was meaning to try this at some point, glad you got it working.

TroyNeubauer commented:
Took another pass!

TroyNeubauer commented:
Btw I added CI for android which builds the whole thing on aarch64 and armv7. Also switched `actions-rs/toolchain@v1` (deprecated) to `dtolnay/rust-toolchain@stable`

See: https://github.com/actions-rs/toolchain/issues/216

kevinmehall commented:
Thanks!

Tested on Linux with a device that supports GET_CONFIGURATION and one that doesn't.

It would be useful to have example code to point to on how to integrate this with the Java APIs on Android, so if you publish code for an app that uses this, please let me know!

TroyNeubauer commented:
Certainly! Getting the permissions setup and opening the device to get the FD is a bit annoying, so an example would be great to have. I'll put up a PR up for this in a bit when I have time

kevinmehall commented on src/enumeration.rs:
```diff
@@ -314,6 +314,9 @@ impl std::fmt::Debug for DeviceInfo {
             s.field("sysfs_path", &self.path);
+        }
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        {
```
I missed this, but @tuna-f1sh [noticed](https://github.com/kevinmehall/nusb/pull/77#issuecomment-2368027477) when trying to rebase on top of it -- what's your reasoning for splitting this to exclude `sysfs_path` here, and similarly not adding `target_os = "android"` to the `DeviceInfo::sysfs_path` getter?

Un-rooted Android doesn't allow sysfs, but everything in `DeviceInfo` comes from sysfs, so `list_devices` isn't expected to work there at all; that's why you added `from_fd`. I was assuming you added `target_os = "android"` in this file for use on rooted devices, or just consistency with Linux, in which case it seems like `sysfs_path` should be included. Alternatively, should we instead  `cfg` out `list_devices` and `DeviceInfo`  entirely?

TroyNeubauer replied:
Good catch. Yeah this is something I missed.

I feel like the correct thing to do would be to add `target_os = "android"` to the getter, `fmt::Debug` impl etc. for consistency between linux and rooted devices. Then add some docs in `list_devices` noting that this will fail on unrooted android, with a link to `wrap_device` like what libusb does: https://libusb.sourceforge.io/api-1.0/group__libusb__lib.html#gga07d4ec54cf575d672ba94c72b3c0de7ca5534030a8299c85555aebfae39161ef7
